File: scripts/t-wrapper/import-manager.ts
/**
 * Import 관리 유틸리티
 */

import traverse from "@babel/traverse";
import * as t from "@babel/types";
import { STRING_CONSTANTS } from "./constants";

/**
 * useTranslation 훅을 생성하는 AST 노드 생성
 */
export function createUseTranslationHook(): t.VariableDeclaration {
  // useTranslation()을 빈 값으로 호출 - 내부적으로 현재 언어 자동 주입
  const hookCall = t.callExpression(
    t.identifier(STRING_CONSTANTS.USE_TRANSLATION),
    []
  );

  return t.variableDeclaration(STRING_CONSTANTS.VARIABLE_KIND, [
    t.variableDeclarator(
      t.objectPattern([
        t.objectProperty(
          t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
          t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
          false,
          true
        ),
      ]),
      hookCall
    ),
  ]);
}

/**
 * AST에 useTranslation import가 필요한지 확인하고 추가
 */
export function addImportIfNeeded(
  ast: t.File,
  translationImportSource: string
): boolean {
  let hasImport = false;

  traverse(ast, {
    ImportDeclaration: (path) => {
      if (path.node.source.value === translationImportSource) {
        const hasUseTranslation = path.node.specifiers.some(
          (spec) =>
            t.isImportSpecifier(spec) &&
            t.isIdentifier(spec.imported) &&
            spec.imported.name === STRING_CONSTANTS.USE_TRANSLATION
        );

        if (!hasUseTranslation) {
          path.node.specifiers.push(
            t.importSpecifier(
              t.identifier(STRING_CONSTANTS.USE_TRANSLATION),
              t.identifier(STRING_CONSTANTS.USE_TRANSLATION)
            )
          );
        }
        hasImport = true;
      }
    },
  });

  if (!hasImport) {
    const importDeclaration = t.importDeclaration(
      [
        t.importSpecifier(
          t.identifier(STRING_CONSTANTS.USE_TRANSLATION),
          t.identifier(STRING_CONSTANTS.USE_TRANSLATION)
        ),
      ],
      t.stringLiteral(translationImportSource)
    );
    ast.program.body.unshift(importDeclaration);
    return true;
  }

  return false;
}

File: scripts/t-wrapper-rust/src/import_manager.rs
/// Import 관리 유틸리티

use crate::constants::StringConstants;
use swc_ecma_ast::*;
use swc_ecma_visit::{VisitMut, VisitMutWith};
use swc_common::DUMMY_SP;

/// useTranslation 훅을 생성하는 AST 노드 생성
/// 
/// TypeScript 버전과 동일한 로직:
/// const { t } = useTranslation();
/// 
/// TODO: SWC AST 노드 생성으로 구현 필요
/// 현재는 문자열로 반환
pub fn create_use_translation_hook() -> String {
    format!(
        "const {{ {} }} = {}();",
        StringConstants::TRANSLATION_FUNCTION,
        StringConstants::USE_TRANSLATION
    )
    // TODO: 실제 AST 노드 생성
    // VariableDeclaration {
    //     kind: VariableKind::Const,
    //     decls: vec![VariableDeclarator {
    //         name: Pat::Object(ObjectPat {
    //             props: vec![ObjectPatProp::KeyValue(KeyValuePatProp {
    //                 key: PropName::Ident(Ident {
    //                     sym: StringConstants::TRANSLATION_FUNCTION.into(),
    //                     ...
    //                 }),
    //                 value: Box::new(Pat::Ident(BindingIdent {
    //                     id: Ident {
    //                         sym: StringConstants::TRANSLATION_FUNCTION.into(),
    //                         ...
    //                     },
    //                     ...
    //                 })),
    //             })],
    //             ...
    //         }),
    //         init: Some(Box::new(Expr::Call(CallExpr {
    //             callee: Callee::Expr(Box::new(Expr::Ident(Ident {
    //                 sym: StringConstants::USE_TRANSLATION.into(),
    //                 ...
    //             }))),
    //             args: vec![],
    //             ...
    //         }))),
    //         ...
    //     }],
    //     ...
    // }
}

/// AST에 useTranslation import가 필요한지 확인하고 추가
/// 
/// TypeScript 버전과 동일한 로직:
/// 1. AST를 traverse하여 ImportDeclaration 확인
/// 2. 같은 소스의 import가 있고 useTranslation이 있으면 false 반환
/// 3. 같은 소스의 import가 있지만 useTranslation이 없으면 specifier 추가
/// 4. import가 없으면 새로 생성
pub fn add_import_if_needed(module: &mut Module, translation_import_source: &str) -> bool {
    let mut has_import = false;
    let mut has_use_translation = false;

    // TODO: SWC AST traverse로 구현
    // module.visit_mut_with(&mut ImportVisitor {
    //     translation_import_source,
    //     has_import: &mut has_import,
    //     has_use_translation: &mut has_use_translation,
    // });

    // 기존 import 확인
    // TODO: Wtf8Atom을 문자열로 변환하는 올바른 방법 찾기
    for stmt in &module.body {
        // TODO: import 확인 로직 구현
        // if let ModuleItem::ModuleDecl(ModuleDecl::Import(import_decl)) = stmt {
        //     // Wtf8Atom 처리 필요
        //     // if import_decl.src.value == translation_import_source {
        //     //     has_import = true;
        //     //     // specifiers 확인
        //     //     for spec in &import_decl.specifiers {
        //     //         if let ImportSpecifier::Named(named) = spec {
        //     //             if named.local.sym == StringConstants::USE_TRANSLATION {
        //     //                 has_use_translation = true;
        //     //             }
        //     //         }
        //     //     }
        //     // }
        // }
        let _ = stmt;
    }

    // useTranslation이 없으면 추가
    if has_import && !has_use_translation {
        // TODO: 기존 import에 specifier 추가
        // if let Some(import_decl) = find_import_declaration(module, translation_import_source) {
        //     import_decl.specifiers.push(ImportSpecifier::Named(NamedImportSpecifier {
        //         local: Ident {
        //             sym: StringConstants::USE_TRANSLATION.into(),
        //             ...
        //         },
        //         imported: None,
        //         ...
        //     }));
        //     return true;
        // }
        return true;
    }

    // import가 아예 없으면 새로 생성
    if !has_import {
        // TODO: 새 import declaration 생성 및 추가
        // let import_decl = ImportDecl {
        //     src: Box::new(Str {
        //         value: translation_import_source.into(),
        //         ...
        //     }),
        //     specifiers: vec![ImportSpecifier::Named(NamedImportSpecifier {
        //         local: Ident {
        //             sym: StringConstants::USE_TRANSLATION.into(),
        //             ...
        //         },
        //         imported: None,
        //         ...
        //     })],
        //     ...
        // };
        // module.body.insert(0, ModuleItem::ModuleDecl(ModuleDecl::Import(import_decl)));
        return true;
    }

    false
}

/// Server translation function import 추가
pub fn add_server_translation_import(
    module: &mut Module,
    translation_import_source: &str,
    server_function_name: &str,
) -> bool {
    // TODO: server translation function import 추가
    // addImportIfNeeded와 유사한 로직이지만 server_function_name을 사용
    let _ = (module, translation_import_source, server_function_name);
    true
}
